fcrepo-search-impl/src/main/java/org/fcrepo/search/impl/DbSearchIndexImpl.java

```diff
@@ -304,154 +535,216 @@ private String convertToSqlLikeWildcard(final String value) {
+                                                        final String txId,
+                                                        final FedoraId fedoraId) {
+        //remove and add type associations for the fedora id.
+        final List<MapSqlParameterSource> parameterSourcesList = new ArrayList<>(rdfTypes.size());
+        final var parameterSource = new MapSqlParameterSource();
+        parameterSource.addValue(TRANSACTION_ID_PARAM, txId);
+        parameterSource.addValue(FEDORA_ID_PARAM, fedoraId.getFullId());
+        jdbcTemplate.update(DELETE_RESOURCE_TYPE_ASSOCIATIONS_IN_TRANSACTION, parameterSource);
+
+        for (final var rdfType : rdfTypes) {
+            final var assocParams = new MapSqlParameterSource();
+            assocParams.addValue(TRANSACTION_ID_PARAM, txId);
+            assocParams.addValue(FEDORA_ID_PARAM, fedoraId.getFullId());
+            assocParams.addValue(RDF_TYPE_URI_PARAM, rdfType.toString());
+            parameterSourcesList.add(assocParams);
         }
+        final MapSqlParameterSource[] psArray = parameterSourcesList.toArray(new MapSqlParameterSource[0]);
+        jdbcTemplate.batchUpdate(INSERT_RDF_TYPE_ASSOC_IN_TRANSACTION, psArray);
+    }
 
-        public String getTypeUri() {
-            return typeUri;
+    private void deleteRdfTypeAssociations(final FedoraId fedoraId) {
+        final var deleteParams = new MapSqlParameterSource();
+        deleteParams.addValue(FEDORA_ID_PARAM, fedoraId.getFullId());
+        jdbcTemplate.update(DELETE_RDF_TYPE_ASSOCIATIONS,
+                deleteParams);
+    }
+
+    private void insertRdfTypeAssociations(final List<URI> rdfTypes, final FedoraId fedoraId) {
+        //add rdf type associations
+
+        final List<MapSqlParameterSource> parameterSourcesList = new ArrayList<>();
+        for (final var rdfType : rdfTypes) {
+            final var assocParams = new MapSqlParameterSource();
+            assocParams.addValue(FEDORA_ID_PARAM, fedoraId.getFullId());
+            assocParams.addValue(RDF_TYPE_URI_PARAM, rdfType.toString());
+            parameterSourcesList.add(assocParams);
         }
+        final MapSqlParameterSource[] psArray = parameterSourcesList.toArray(new MapSqlParameterSource[0]);
+        jdbcTemplate.batchUpdate(INSERT_RDF_TYPE_ASSOC, psArray);
     }
 
+
+    @TransactionalWithRetry
     @Override
-    public void removeFromIndex(final FedoraId fedoraId) {
-        try {
-            final var params = new MapSqlParameterSource();
-            params.addValue(FEDORA_ID_PARAM, fedoraId.getFullId());
-            jdbcTemplate.update(DELETE_FROM_INDEX_SQL, params);
-        } catch (final DataAccessException ex) {
-            throw new RepositoryRuntimeException("Failed to delete search index entry for " + fedoraId.getFullId());
-        }
+    public void removeFromIndex(final Transaction transaction, final FedoraId fedoraId) {
+        transaction.doInTx(() -> {
+            if (transaction.isOpenLongRunning()) {
```

Comment:
This test is not quite right because you want it to run the long-running tx branch when it is a long-running tx, regardless of the state of the tx.

Reply:
👍